### PR TITLE
Pract.02: Add 04-07 | Kiril Georgiev

### PR DESCRIPTION
**Pull request by KirilGeorgiev247**
Kiril Georgiev
SI 1 8MI0600268

### CODE REVIEW

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	unsigned int b;
+};
+
+struct Relation
+{
+	unsigned int pairsCount;
+	Pair pairs[RELATION_PAIRS_MAX_COUNT];
+};
+
+void convertStringToInt(const char* str, int& num, const int numLength)
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp:21]: Do not use & with primitive data types

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	int firstNumAsInt = 0;
+	int secondNumAsInt = 0;
+
+	convertStringToInt(firstNum, firstNumAsInt, firstNumIndex);
+	convertStringToInt(secondNum, secondNumAsInt, secondNumIndex);
+
+	pair.a = (unsigned int)firstNumAsInt;
+	pair.b = (unsigned int)secondNumAsInt;
+
+	file.seekg(prevPos);
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp:77]: You have written so many redundant lines of code 
As I can see the way you are writing to the file you can just write
while(!file.eof()){
int a, b;
file>>a>>b;
pair.a=a'
pair.b=b;

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	while (line[0] != '\0')
+	{
+		user.password[index++] = line[0];
+		line++;
+	}
+	user.password[index] = '\0';
+}
+
+// returns true if their symbols are matching
+bool compareStrings(const char* firstString, const char* secondString)
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:67]: what happens when secondString size is less than the size of firstString

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+const int COMMAND_MAX_LENGTH = 31;
+const char* REGISTER_COMMAND = "register";
+const char* LOGIN_COMMAND = "login";
+const char* FILE_NAME = "SystemOfUsers.txt";
+
+struct User
+{
+	char* name = new char[USER_PROPERTY_MAX_LENGTH];
+	char* email = new char[USER_PROPERTY_MAX_LENGTH];
+	char* password = new char[USER_PROPERTY_MAX_LENGTH];
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:17]: You do not free the allocated memory

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	{
+		num += ((int)str[i]) * helpingNum;
+		helpingNum *= 10;
+	}
+}
+
+Pair* CreatePair(unsigned int a, unsigned int b)
+{
+	Pair pair{ a,b };
+	return &pair;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp:34]: BIG mistake. If this was at the exam, you would receive 0 points for the whole exam.

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+{
+	if (relation.pairsCount < RELATION_PAIRS_MAX_COUNT)
+	{
+		relation.pairs[relation.pairsCount++] = pair;
+	}
+}
+
+int main()
+{
+	Pair pair1 = *CreatePair(4, 6);
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp:95]: big mistake

**Review by khatoni — CHANGES_REQUESTED** [review]

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+				}
+			}
+			else
+			{
+				std::cout << "A user with this email does not exists!";
+			}
+			delete[] currUser;
+		}
+		myFile.close();
+	}
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:187]: You do not free the dynamic memory in system

**Review by khatoni — CHANGES_REQUESTED** [review]

**Comment by Justsvetoslavov** [comment]
Check uploaded solutions.
You can re-open the PR or open a new one if you have better solution or questions.

**Pull request closed without merging**